### PR TITLE
MGDAPI-3902 implement flag for disabling syncing of the replicas

### PR DESCRIPTION
**Pull request by MStokluska**
## Related GH Issue
Closes: https://github.com/keycloak/keycloak-operator/issues/549
Jira: https://issues.redhat.com/browse/MGDAPI-3902

## Additional Information
For RHOAM, we would like to introduce a flag in the kc CR to disable syncing replicas between keycloak CR and stateful sets.

## Verification Steps
- create new kc instance with the flag `kc.Spec.DisableReplicasSyncing` set to true.
- on creation of the instance verify that the replicas count is at 0
- update kc instances to 2 via kc cr
- verify that the stateful set remains at 0

## Checklist:
- [ ] Automated Tests
- [ ] [Documentation](https://github.com/keycloak/keycloak-documentation) changes if necessary
- [ ] (If required) Discussed on [Keycloak DEV](https://groups.google.com/forum/#!forum/keycloak-dev)

### CODE REVIEW

File: pkg/apis/keycloak/v1alpha1/keycloak_types.go [hunk 1]
```diff
@@ -87,6 +87,10 @@ type KeycloakSpec struct {
 	// objects and users will have to create them manually, if needed.
 	// +optional
 	DisableMonitoringServices bool `json:"DisableDefaultServiceMonitor,omitempty"`
+	// Specify whether disabling the syncing of instances from the Keycloak CR to the statefulset replicas
+	// should be enabled or disabled. Defaults to false.
```
Comment (vmuzikar) [inline pkg/apis/keycloak/v1alpha1/keycloak_types.go:91]: Maybe should we mention it's particularly useful for HPAs?

File: pkg/model/rhsso_deployment.go [hunk 2]
```diff
@@ -239,7 +239,9 @@ func RHSSODeploymentReconciled(cr *v1alpha1.Keycloak, currentState *v13.Stateful
 	reconciled.Spec.Template.ObjectMeta.Annotations = AddPodAnnotations(cr, reconciled.Spec.Template.ObjectMeta.Annotations)
 
 	reconciled.ResourceVersion = currentState.ResourceVersion
-	reconciled.Spec.Replicas = SanitizeNumberOfReplicas(cr.Spec.Instances, false)
+	if cr.Spec.DisableReplicasSyncing != true {
+		reconciled.Spec.Replicas = SanitizeNumberOfReplicas(cr.Spec.Instances, false)
+	}
```
Comment (vmuzikar) [inline pkg/model/rhsso_deployment.go:244]: Could you please do a similar implementation in `keycloak_deployment` as well?

**Review by vmuzikar — CHANGES_REQUESTED** [review]
@MStokluska Thank you for your contribution!

Added a few comments and approved the CI.

File: pkg/apis/keycloak/v1alpha1/keycloak_types.go (hunk 1 above)
Comment (MStokluska) [reply pkg/apis/keycloak/v1alpha1/keycloak_types.go:91]: Thanks @vmuzikar! good idea

File: pkg/model/rhsso_deployment.go (hunk 2 above)
Comment (MStokluska) [reply pkg/model/rhsso_deployment.go:244]: Absolutely! thank you

File: pkg/model/rhsso_deployment_test.go [hunk 3]
```diff
@@ -31,3 +35,49 @@ func TestRHSSODeployment_testAffinityExperimental(t *testing.T) {
...
+
+func TestRHSSODeploymentReconciled_testDisableRHSSOReplicasSyncingFalse(t *testing.T) {
+	testDisableRHSSOReplicasSyncingFalse(t, RHSSODeployment, RHSSODeploymentReconciled)
+}
+
+func TestRHSSODeploymentReconciled_testDisableRHSSOReplicasSyncingTrue(t *testing.T) {
+	testDisableRHSSOReplicasSyncingTrue(t, RHSSODeployment, RHSSODeploymentReconciled)
+}
+
+func testDisableRHSSOReplicasSyncingFalse(t *testing.T, deploymentFunction createDeploymentStatefulSet, deploymentFunction2 reconcileRHSSODeployment) {
```
Comment (vmuzikar) [inline pkg/model/rhsso_deployment_test.go:47]: Just to align with other tests in this file, could you please move this to `keycloak_deployment_test`?

File: pkg/model/rhsso_deployment_test.go [hunk 4]
```diff
@@ -31,3 +35,49 @@ func TestRHSSODeployment_testAffinityExperimental(t *testing.T) {
...
+
+	// when
+	statefulSet.Spec.Replicas = &[]int32{4}[0]
+	replicasCount := deploymentFunction2(cr, statefulSet, dbSecret, nil).Spec.Replicas
+
+	//then
+	assert.Equal(t, int32(2), int32(*replicasCount))
+}
+
+func testDisableRHSSOReplicasSyncingTrue(t *testing.T, deploymentFunction createDeploymentStatefulSet, deploymentFunction2 reconcileRHSSODeployment) {
```
Comment (vmuzikar) [inline pkg/model/rhsso_deployment_test.go:66]: Just to align with other tests in this file, could you please move this to `keycloak_deployment_test`?

**Review by vmuzikar — CHANGES_REQUESTED** [review]
@MStokluska Thanks for the update! Added two (last :)) comments. On top of that there are some linter errors. Could you please check it out?

File: pkg/model/rhsso_deployment_test.go (hunk 4 above)
Comment (MStokluska) [reply pkg/model/rhsso_deployment_test.go]: Thank you, done!

File: pkg/model/rhsso_deployment_test.go (hunk 3 above)
Comment (MStokluska) [reply pkg/model/rhsso_deployment_test.go]: Thank you @vmuzikar, done!

File: pkg/model/keycloak_deployment_test.go [hunk 5]
```diff
@@ -492,3 +501,77 @@ func testServiceAccountDefault(t *testing.T, deploymentFunction createDeployment
 
 	assert.Equal(t, "default", serviceAccountName)
 }
+
+func testDisableDeploymentReplicasSyncingFalse(t *testing.T, deploymentFunction createDeploymentStatefulSet, deploymentFunction2 reconcileRHSSODeployment) {
```
Comment (vmuzikar) [inline pkg/model/keycloak_deployment_test.go:505]: After the latest changes, I noticed we have two pairs of basically identical test functions here:
* `testDisableDeploymentReplicasSyncingFalse` and `testDisableRHSSOReplicasSyncingFalse`
* `testDisableDeploymentReplicasSyncingTrue` and `testDisableRHSSOReplicasSyncingTrue`

Is there a reason for this? Maybe I'm just missing some context here. :) Could we possibly keep just `testDisableDeploymentReplicasSyncingFalse` and `testDisableDeploymentReplicasSyncingTrue` and use them both in `keycloak_deployment_test` and `rhsso_deployment_test`?

Other than that LGTM. 👍

Comment (MStokluska) [reply pkg/model/keycloak_deployment_test.go:505]: good catch! Thanks @vmuzikar 
Updated + squashed the commits. 
Thank you

**Review by vmuzikar — APPROVED** [review]
@MStokluska LGTM, thank you!

**Pull request merged**